tensorflow_serving/batching/batching_util.cc

```diff
@@ -0,0 +1,229 @@
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow_serving/batching/batching_util.h"
+
+#include <string>
+
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/lib/core/errors.h"
+
+
+namespace tensorflow {
+namespace serving {
+
+// Padding for one dimension of a tensor.
+struct OneDimPadding {
+  int64 first;  // pad before
+  int64 second;  // pad after
+};
+
+// Constructs array of paddings, where:
+// paddings[i].first - number of objects to add before elements in dimension i
+// of given tensor,
+// paddings[i].second - number of objects to add after elements in dimension i.
+// This padding signature is used when performing internal padding with Eigen.
+//
+// When building paddings it is assumed that tensor needs to be padded
+// after each dimension, so its shape matches max_dim_sizes,
+// First entry of max_dim_sizes, which is maximum size of zeroth dimension,
+// is ignored, because we don't perform padding in that dimension.
+template<int num_dims>
+std::array<OneDimPadding, num_dims> CreatePadding(Tensor tensor,
+        const std::vector<int>& max_dim_sizes) {
+  std::array<OneDimPadding, num_dims> padding;
+  for (unsigned int i = 0; i < max_dim_sizes.size(); ++i) {
+    if (i > 0 && max_dim_sizes[i] - tensor.dim_size(i) > 0) {
+      padding[i] = {0, max_dim_sizes[i] - tensor.dim_size(i)};
+    } else {
+      padding[i] = {0, 0};
+    }
+  }
+  return padding;
+}
+
+// Functor, which performs padding of given input tensor
+// using specified padding signature.
+// For example, given tensor of shape [1, 2, 3] and padding signature
+// [[0, 0], [0, 2], [2, 2]]
+// functor produces padded_tensor of shape [1, 4, 7].
+template <typename T, int num_dims>
+struct PadTensor {
+  Status operator()(Tensor input,
+      const std::array<OneDimPadding, num_dims>& padding,
+      Tensor* output) {
+    TensorShape output_shape;
+    for (int d = 0; d < num_dims; ++d) {
+      // Pad before existing elements.
+      const int32 before_d = padding[d].first;
+      // Pad after existing elements.
+      const int32 after_d = padding[d].second;
+      output_shape.AddDim(before_d + input.dim_size(d) + after_d);
+    }
+    if (output_shape.num_elements() == input.NumElements()) {
+      auto result = output->CopyFrom(input, output_shape);
+      if (!result) {
+        return errors::Internal("Couldn't create output.");
+      }
+      return Status::OK();
+    }
+    *output = Tensor(input.dtype(), output_shape);
+    typename TTypes<T, num_dims>::Tensor inputs = input.tensor<T, num_dims>();
+    T pad_value(input.flat<T>()(0));  // using existing values in padding
+    output->tensor<T, num_dims>() = inputs.pad(padding, pad_value);
+    return Status::OK();
+  }
+};
+
+// functor specialization for scalars:
```

Comment:
Do we need this? Does it even make sense to batch scalars? I'm not sure if the rest of the batching codebase supports scalars, so there may be no point in supporting them here. (But I could be wrong.)